rewrite-java/src/main/java/org/openrewrite/java/JavaTemplate.java

```diff
@@ -247,13 +246,85 @@ public J visitMethodDeclaration(J.MethodDeclaration method, Integer p) {
+                                    }
+                                    if(!(decl.getTypeExpression() instanceof J.Identifier || decl.getTypeExpression() instanceof J.Primitive)) {
+                                        throw new IllegalArgumentException(
+                                                "Only an Identifier or a Primitive can be supplied as the type expression of a parameter " +
+                                                        "appearing in a method declaration's parameter list: " + parameter.print());
+                                    }
+                                    J.VariableDeclarations.NamedVariable namedVariable = decl.getVariables().get(0);
+                                    paramNames.add(namedVariable.getSimpleName());
+                                    if(namedVariable.getType() == null) {
+                                        // null if the type of the argument is a generic type parameter
+                                        // Try to find an appropriate type from the method itself
+                                        assert decl.getTypeExpression() instanceof J.Identifier;
+                                        J.Identifier declTypeIdent = (J.Identifier) decl.getTypeExpression();
+                                        String typeParameterName = declTypeIdent.getSimpleName();
+                                        for(J.TypeParameter typeParameter : method.getTypeParameters()) {
+                                            J.Identifier typeParamIdent = (J.Identifier) typeParameter.getName();
+                                            if(typeParamIdent.getSimpleName().equals(typeParameterName)) {
+                                                List<TypeTree> bounds = typeParameter.getBounds();
+                                                JavaType.FullyQualified bound;
+                                                if(bounds.size() == 0) {
+                                                    bound = JavaType.Class.OBJECT;
+                                                } else {
+                                                    bound = (JavaType.FullyQualified) bounds.get(0);
+                                                }
+                                                JavaType.GenericTypeVariable genericType = new JavaType.GenericTypeVariable(
+                                                        typeParamIdent.getSimpleName(),
+                                                        bound);
+
+                                                paramTypes.add(genericType);
+                                            }
+                                        }
+                                    } else {
+                                        paramTypes.add(namedVariable.getType());
+                                    }
+                                }
+
+                                newType = newType.withParamNames(paramNames)
```

Comment:
Wondering if we should have wither methods on JavaType.Method? We probably should be using `JavaType.Method.build()` here to make sure we go through the flyweights.

Reply:
Good point. I'll use JavaType.Method.build instead and remove the `@With`s from JavaType.Method.